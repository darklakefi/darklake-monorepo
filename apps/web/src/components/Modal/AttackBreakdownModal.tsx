import Modal, { ModalProps } from "@/components/Modal";
import { MevAttack, MevAttackSwapType } from "@/types/Mev";
import { format } from "date-fns";
import { formatMoney, formatPercentage } from "@/utils/number";
import { cn, truncate } from "@/utils/common";

interface AttackBreakdownModalProps extends ModalProps {
  mevAttack?: MevAttack;
}

export default function AttackBreakdownModal(props: AttackBreakdownModalProps) {
  const { mevAttack } = props;

  if (!mevAttack?.transactions || !mevAttack?.solAmount) {
    return (
      <Modal title="Attack breakdown" {...props}>
        <div className="lg:w-[960px] p-5 text-brand-30">
          <p>Unable to load attack details. Please try again later.</p>
        </div>
      </Modal>
    );
  }

  const isVictimBuy = mevAttack.swapType === MevAttackSwapType.BUY;
  const solAmountLostFormatted = formatMoney(mevAttack.solAmount.lost);
  const solAmountSentFormatted = formatMoney(mevAttack.solAmount.sent);

  return (
    <Modal title="Attack breakdown" {...props}>
      <div className="lg:w-[960px]">
        <div className="uppercase text-lg leading-6 tracking-normal text-brand-30 mb-6">
          <p>Token: {mevAttack.tokenName}</p>
          <p>Date: {format(mevAttack.timestamp, "yyyy-MM-dd hh:mm OOO")}</p>
        </div>
        <div className="flex flex-row justify-between items-center p-5 bg-brand-60 mb-6">
          {[
            { title: "Total lost", value: `${solAmountLostFormatted} SOL` },
            {
              title: "Extracted",
              value: `${formatPercentage((mevAttack.solAmount.lost / mevAttack.solAmount.sent) * 100)}%`,
            },
            { title: "TX size", value: `${solAmountSentFormatted} SOL` },
          ].map((row) => (
            <div key={row.title} className="uppercase">
              <p className="text-3xl leading-9 text-brand-20">{row.value}</p>
              <p className="text-lg leading-6 tracking-normal text-brand-30">{row.title}</p>
            </div>
          ))}
        </div>
        <div className="relative">
          <span className="w-px h-full absolute left-[11.5px] max-md:hidden top-0 bg-brand-50" />
          {[
            {
              title: "Frontrun Attack",
              iconClassName: "hn-exclamation-triangle-solid",
              details:
                `Attacker ${isVictimBuy ? "buys" : "sells"} ${mevAttack.tokenName} ` +
                `tokens to drive ${isVictimBuy ? "up" : "down"} price`,
              transaction: mevAttack.transactions.frontRun,
            },
            {
              title: "Victim Transaction",
              iconClassName: "hn-refresh-solid",
              details:
                `Swap: ${isVictimBuy ? `${solAmountSentFormatted} SOL` : mevAttack.tokenName} ` +
                `→ ${isVictimBuy ? mevAttack.tokenName : `${solAmountSentFormatted} SOL`} ` +
                `– executed at ${isVictimBuy ? "inflated" : "dropped"} price`,
              transaction: mevAttack.transactions.victim,
            },
            {
              title: "Backrun Attack",
              iconClassName: "hn-exclamation-triangle-solid",
              details:
                `Attacker ${isVictimBuy ? "sells" : "buys"} ${mevAttack.tokenName} ` +
                `tokens at ${isVictimBuy ? "peak" : "floor"} price ` +
                `– value extracted: ${solAmountLostFormatted} SOL`,
              transaction: mevAttack.transactions.frontRun,
            },
          ].map((row) => (
            <div
              key={`${row.title}-${row.transaction.address}${row.transaction.address}-${row.transaction.signature}`}
              className={cn(
                "md:flex flex-row items-center w-full gap-x-2 md:mb-6 relative",
                "uppercase text-lg leading-5",
                "group",
              )}
            >
              <span className="hidden md:group-first-of-type:block absolute top-0 left-0 bg-brand-70 h-1/2 w-full" />
              <span className="hidden md:group-last-of-type:block absolute bottom-0 left-0 bg-brand-70 h-1/2 w-full" />
              <div
                className={cn(
                  "text-brand-20 md:w-[264px]",
                  "flex flex-row items-center justify-center py-5 bg-brand-70 relative",
                )}
              >
                <i className={cn("text-[22px] hn mr-4", row.iconClassName)} />
                <p>{row.title}</p>
                <span className="max-md:hidden ml-1 h-px w-20 bg-brand-50 flex-1" />
              </div>
              <div className=" text-brand-30 border border-brand-60 p-5 flex-1 relative">
                <p className="text-brand-20">{row.details}</p>
                <p>{truncate(row.transaction.signature)}</p>
                <a
                  href={`${process.env.NEXT_PUBLIC_EXPLORER_URL}/tx/${row.transaction.signature}`}
                  className="underline"
                  title="View Transaction"
                  target="_blank"
                >
                  View TX
                </a>
              </div>
              <span className="block group-last-of-type:hidden md:hidden w-px mt-4 h-8 mx-auto bg-brand-50" />
            </div>
          ))}
        </div>
      </div>
    </Modal>
  );
}
