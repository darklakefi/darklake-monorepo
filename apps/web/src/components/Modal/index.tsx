"use client";

import { PropsWithChildren, useState } from "react";
import { Dialog, DialogBackdrop, DialogPanel } from "@headlessui/react";
import { cn } from "@/utils/common";

export interface ModalProps {
  isOpen: boolean;
  title?: string;
  onClose?: () => void;
  contentWrapperClassName?: string;
  children?: React.ReactNode;
}

export default function Modal({
  title,
  children,
  isOpen,
  onClose,
  contentWrapperClassName,
}: PropsWithChildren<ModalProps>) {
  const [forceHide, setForceHide] = useState(false);

  const onModalClose = () => {
    if (onClose) {
      onClose();
    } else {
      setForceHide(true);
    }
  };

  return (
    <Dialog as="div" className="relative" open={!forceHide && isOpen} onClose={onModalClose}>
      <DialogBackdrop className="fixed inset-0 bg-modal-backdrop" />
      <div className="flex flex-col fixed inset-0 w-screen items-center justify-center">
        <DialogPanel>
          <div className="py-4 pl-4 pr-7">
            <div
              className={cn(
                "max-h-[80vh] lg:max-h-none overflow-auto",
                "p-[16px] md:p-[40px] bg-brand-70",
                "shadow-[12px_12px_0px_0px] shadow-brand-60 relative",
                contentWrapperClassName,
              )}
            >
              <div
                className={cn(
                  "flex flex-row items-center mb-4",
                  title?.length ? "border-b border-b-brand-50 pb-3 justify-between" : "justify-end",
                )}
              >
                {!!title?.length && <p className="font-primary text-3xl leading-3xl text-brand-20">{title}</p>}
                <button type="button" onClick={onModalClose} className="hover-with-active flex items-center">
                  <i className="hn hn-times text-brand-30 text-xl leading-5" />
                </button>
              </div>
              <div className="font-secondary text-lg leading-6 text-brand-30">{children}</div>
            </div>
          </div>
        </DialogPanel>
      </div>
    </Dialog>
  );
}
