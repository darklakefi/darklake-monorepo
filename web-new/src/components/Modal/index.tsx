"use client";

import { PropsWithChildren, useState } from "react";
import clsx from "clsx";
import Image from "next/image";
import { Dialog, DialogBackdrop, DialogPanel } from "@headlessui/react";
import iconClose from "../../../public/images/icon-close.png";

export interface ModalProps {
  isOpen: boolean;
  title?: string;
  onClose?: () => void;
  contentWrapperClassName?: string;
}

export default function Modal({
  title,
  children,
  isOpen,
  onClose,
  contentWrapperClassName,
}: PropsWithChildren<ModalProps>) {
  const [forceHide, setForceHide] = useState(false);

  const onModalClose = () => {
    if (onClose) {
      onClose();
    } else {
      setForceHide(true);
    }
  };

  return (
    <Dialog as="div" className="relative" open={!forceHide && isOpen} onClose={onModalClose}>
      <DialogBackdrop className="fixed inset-0 bg-modal-backdrop" />
      <div className="flex flex-col fixed inset-0 w-screen items-center justify-center">
        <DialogPanel>
          <div className="py-[16px] pl-[16px] pr-[28px]">
            <div
              className={clsx(
                "max-h-[80vh] lg:max-h-none overflow-auto",
                "p-[16px] md:p-[40px] bg-brand-70",
                "shadow-[12px_12px_0px_0px] shadow-brand-60 relative",
                contentWrapperClassName,
              )}
            >
              <div
                className={clsx(
                  "flex flex-row items-center mb-[16px]",
                  title?.length ? "border-b border-b-brand-50 pb-[12px] justify-between" : "justify-end",
                )}
              >
                {!!title?.length && <p className="text-heading-1 text-brand-20">{title}</p>}
                <button type="button" onClick={onModalClose} className="active-hover-opacity">
                  <Image src={iconClose} alt="modal close" width={20} height={20} />
                </button>
              </div>
              <div className="text-body-2 text-brand-30">
                {/* @ts-expect-error: React 18 vs 19 type issue */}
                {children}
              </div>
            </div>
          </div>
        </DialogPanel>
      </div>
    </Dialog>
  );
}
