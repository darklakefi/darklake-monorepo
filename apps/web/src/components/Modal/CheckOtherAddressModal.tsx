import Modal, { ModalProps } from "@/components/Modal";
import { LocalStorage } from "@/constants/storage";
import useLocalStorage from "@/hooks/useLocalStorage";
import AddressMevLookup from "../AddressMevLookup";

export default function CheckOtherAddressModal(props: ModalProps) {
  const [lookupAddress] = useLocalStorage<string | null>(LocalStorage.LOOKUP_ADDRESS, null);
  return (
    <Modal title="Check other address" {...props}>
      <div className="flex flex-col gap-8 lg:w-[720px]">
        <AddressMevLookup onSubmit={props.onClose} />
        <div className="flex flex-col gap-1">
          <div className="uppercase text-body-2 text-brand-30">address currently added:</div>
          <div className="text-body-2 text-brand-20 break-all">{lookupAddress}</div>
        </div>
      </div>
    </Modal>
  );
}
