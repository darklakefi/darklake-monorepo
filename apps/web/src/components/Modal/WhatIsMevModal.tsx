import Modal, { ModalProps } from "@/components/Modal";

export default function WhatIsMevModal(props: ModalProps) {
  return (
    <Modal title="What is MEV?" {...props}>
      <div className="md:w-[430px]">
        <p className="uppercase text-body-2">
          MEV (MAXIMUM EXTRACTABLE VALUE) REFERS TO THE PROFITS THAT MINERS, VALIDATORS, OR ANYONE CONTROLLING
          TRANSACTION ORDERING CAN EXTRACT BY REORDERING, INCLUDING, OR EXCLUDING TRANSACTIONS WITHIN A BLOCK.
          <br />
          <br />
          ONE OF THE MOST COMMON FORMS OF MEV IS THE SANDWICH ATTACK. THIS HAPPENS WHEN A MALICIOUS ACTOR PLACES A TRADE
          RIGHT BEFORE AND AFTER A USER&#39;S TRADE TO MANIPULATE THE PRICE, EXTRACTING PROFIT IN THE PROCESS.
          <br />
          <br />
          In this tool, we calculate the MEV loss specifically from Sandwich Attacks on Raydium and Pump.fun.
        </p>
      </div>
    </Modal>
  );
}
