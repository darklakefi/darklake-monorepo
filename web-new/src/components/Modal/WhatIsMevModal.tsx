import Modal, { ModalProps } from "@/components/Modal";

export default function WhatIsMevModal(props: ModalProps) {
  return (
    <Modal title="What is MEV?" {...props}>
      <p className="uppercase">
        MEV (Maximum Extractable Value) refers to the profits that miners, validators, or anyone controlling transaction
        ordering can extract by reordering, including, or excluding transactions within a block.
        <br />
        <br />
        One of the most common forms of MEV is the <span className="text-brand-20">Sandwich Attack</span>. This happens
        when a malicious actor places a trade right before and after a user&#39;s trade to manipulate the price,
        extracting profit in the process.
        <br />
        <br />
        In this tool, we calculate the MEV loss specifically from Sandwich Attacks on Raydium and Pump.fun.
      </p>
    </Modal>
  );
}
