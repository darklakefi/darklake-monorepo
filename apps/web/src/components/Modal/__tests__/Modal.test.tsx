import { act, render, RenderResult } from "@testing-library/react";
import Modal from "@/components/Modal";

describe("<Modal />", () => {
  it("does not display content when closed", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(
        <Modal title="Some modal title" isOpen={false}>
          Some content.
        </Modal>,
      ));
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("displays content successfully when open", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(
        <Modal title="Some modal title" isOpen>
          Some content.
        </Modal>,
      ));
    });

    expect(baseElement).toMatchSnapshot();
  });
});
