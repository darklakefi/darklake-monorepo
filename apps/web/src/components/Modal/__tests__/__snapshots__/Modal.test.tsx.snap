// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> displays content successfully when open 1`] = `
<body>
  <div />
  <div
    id="headlessui-portal-root"
  >
    <div
      data-headlessui-portal=""
    >
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
      <div>
        <div
          aria-modal="true"
          class="relative"
          data-headlessui-state="open"
          data-open=""
          id="headlessui-dialog-:r0:"
          role="dialog"
          tabindex="-1"
        >
          <div
            aria-hidden="true"
            class="fixed inset-0 bg-modal-backdrop"
            data-headlessui-state="open"
            data-open=""
          />
          <div
            class="flex flex-col fixed inset-0 w-screen items-center justify-center"
          >
            <div
              data-headlessui-state="open"
              data-open=""
              id="headlessui-dialog-panel-:r7:"
            >
              <div
                class="py-4 pl-4 pr-7"
              >
                <div
                  class="max-h-[80vh] lg:max-h-none overflow-auto p-4 md:p-10 bg-brand-70 shadow-3xl shadow-brand-80 relative"
                >
                  <div
                    class="flex flex-row items-center mb-4 border-b border-b-brand-50 pb-3 justify-between"
                  >
                    <p
                      class="font-primary text-3xl leading-3xl text-brand-20"
                    >
                      Some modal title
                    </p>
                    <button
                      class="hover-with-active flex items-center"
                      type="button"
                    >
                      <i
                        class="hn hn-times text-brand-30 text-xl leading-5"
                      />
                    </button>
                  </div>
                  <div
                    class="font-secondary text-lg leading-6 tracking-normal text-brand-30"
                  >
                    Some content.
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`<Modal /> does not display content when closed 1`] = `
<body>
  <div />
</body>
`;
